OrbitLinuxTracing/ContextSwitchAndThreadStateVisitor.cpp

```diff
@@ -0,0 +1,211 @@
+
+#include <algorithm>
+#include <utility>
+
+#include "OrbitBase/Logging.h"
+#include "OrbitLinuxTracing/TracerListener.h"
+
+namespace LinuxTracing {
+
+using orbit_grpc_protos::SchedulingSlice;
+using orbit_grpc_protos::ThreadStateSlice;
+
+void ContextSwitchAndThreadStateVisitor::ProcessInitialTidToPidAssociation(pid_t tid, pid_t pid) {
+  bool new_insertion = tid_to_pid_association_.insert_or_assign(tid, pid).second;
+  if (!new_insertion) {
+    ERROR("Overwriting previous pid for tid %d with initial pid %d", tid, pid);
+  }
+}
+
+void ContextSwitchAndThreadStateVisitor::visit(ForkPerfEvent* event) {
+  pid_t pid = event->GetPid();
+  pid_t tid = event->GetTid();
+  bool new_insertion = tid_to_pid_association_.insert_or_assign(tid, pid).second;
+  if (!new_insertion) {
+    ERROR("Overwriting previous pid for tid %d with pid %d from PERF_RECORD_FORK", tid, pid);
+  }
+}
+
+bool ContextSwitchAndThreadStateVisitor::TidMatchesPidFilter(pid_t tid) {
+  if (thread_state_pid_filter_ == kPidFilterNoThreadState) {
+    return false;
+  }
+
+  auto tid_to_pid_it = tid_to_pid_association_.find(tid);
+  if (tid_to_pid_it == tid_to_pid_association_.end()) {
+    return false;
+  }
+
+  return tid_to_pid_it->second == thread_state_pid_filter_;
+}
+
+std::optional<pid_t> ContextSwitchAndThreadStateVisitor::GetPidOfTid(pid_t tid) {
+  auto tid_to_pid_it = tid_to_pid_association_.find(tid);
+  if (tid_to_pid_it == tid_to_pid_association_.end()) {
+    return std::nullopt;
+  }
+  return tid_to_pid_it->second;
+}
+
+void ContextSwitchAndThreadStateVisitor::ProcessInitialState(uint64_t timestamp_ns, pid_t tid,
+                                                             char state_char) {
+  if (!TidMatchesPidFilter(tid)) {
+    return;
+  }
+
+  std::optional<ThreadStateSlice::ThreadState> initial_state = GetThreadStateFromChar(state_char);
+  if (!initial_state.has_value()) {
+    ERROR("Parsing thread state char '%c' for tid %d", state_char, tid);
+    return;
+  }
+  state_manager_.OnInitialState(timestamp_ns, tid, initial_state.value());
+}
+
+void ContextSwitchAndThreadStateVisitor::visit(TaskNewtaskPerfEvent* event) {
+  if (!TidMatchesPidFilter(event->GetTid())) {
+    return;
+  }
+  state_manager_.OnNewTask(event->GetTimestamp(), event->GetTid());
+}
+
+void ContextSwitchAndThreadStateVisitor::visit(SchedSwitchPerfEvent* event) {
+  // Note that context switches with tid 0 are associated with idle CPU, so we never consider them.
+
+  // Process the context switch out for scheduling slices.
+  if (event->GetPrevTid() != 0) {
+    // TracepointPerfEvent::ring_buffer_record.sample_id.pid (which doesn't come from the tracepoint
+    // data, but from the generic field of the PERF_RECORD_SAMPLE) is the pid of the process that
+    // the thread being switched out belongs to. But when the switch out is caused by the thread
+    // exiting, it has value -1. In such cases, use the association between tid and pid that we keep
```

Comment:
It's very annoying, otherwise for scheduling slices we would always have the pid from the switch out. Anyway, I guess it's just how the kernel handles things on exit, I haven't looked at the specifics as the most likely scenario here is that we have to accept it.